core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -4175,98 +4179,491 @@ private RelNode convertUpdate(SqlUpdate call) {
+        Arrays.asList(rowHasUpdateCondition, isMatch));
+
+    RexNode rowHasDeleteCondition;
+
+    if (deleteConds.size() > 1) {
+      rowHasDeleteCondition = relBuilder.getRexBuilder().makeCall(SqlStdOperatorTable.OR,
+          matchCaseNodes.getKey());
+    } else if (deleteConds.size() == 1) {
+      rowHasDeleteCondition = matchCaseNodes.getKey().get(0);
+    } else {
+      rowHasDeleteCondition = relBuilder.getRexBuilder().makeLiteral(false);
+    }
+
+    RexNode isDeleteRow = relBuilder.getRexBuilder().makeCall(SqlStdOperatorTable.AND,
+        Arrays.asList(rowHasDeleteCondition, isMatch));
+
+
+    RexNode isMatchedRow = relBuilder.getRexBuilder().makeCall(
+        SqlStdOperatorTable.OR, Arrays.asList(isUpdateRow, isDeleteRow));
+
+
+    RexNode rowHasInsertCondition;
+    if (notMatchedCaseNodes.getKey().size() > 1) {
+      rowHasInsertCondition = relBuilder.getRexBuilder().makeCall(SqlStdOperatorTable.OR,
+          notMatchedCaseNodes.getKey());
+    } else if (notMatchedCaseNodes.getKey().size() == 1) {
+      rowHasInsertCondition = notMatchedCaseNodes.getKey().get(0);
+    } else {
+      rowHasInsertCondition = relBuilder.getRexBuilder().makeLiteral(false);
     }
 
-    relBuilder.push(join)
-        .project(projects);
+    RexNode isInsertRow = relBuilder.getRexBuilder().makeCall(
+        SqlStdOperatorTable.AND, Arrays.asList(rowHasInsertCondition, isNotMatched));
+
+
+
+    List<RexNode> finalProjects = new ArrayList<>();
+
+    // NOTE: there is a lot of repeated case logic. We should make sure that the
+    // appropriate Rex Nodes are actually being cached/reused when possible
+    for (int colIdx = 0; colIdx < targetTable.getRowType().getFieldCount(); colIdx++) {
+
+      RexNode colNullLiteral = relBuilder.getRexBuilder()
+          .makeNullLiteral(targetTable.getRowType().getFieldList().get(colIdx).getType());
+
+      // Default the update and insert expressions to be NULL literals.
+      // They will be initialized to the appropriate values if we have a matched/insert expression
+      RexNode matchedColExpr = colNullLiteral;
+      RexNode insertColExpr = colNullLiteral;
+
+      // Calculate the matched case (if we have any updates or deletes)
+      if (matchCaseNodes.getKey().size() > 0) {
+        List<RexNode> updateCaseArgs = new ArrayList<>();
+        List<RexNode> curUpdateColExprList = matchCaseNodes.getValue().get(colIdx);
+        for (int updateCondIdx = 0;
+             updateCondIdx < matchCaseNodes.getKey().size(); updateCondIdx++) {
+          // Case operands should are organized as IF, THEN, IF, THEN... ELSE
+          // So, add the condition, followed
+          // by the expression for that column if the condition is True
+          updateCaseArgs.add(matchCaseNodes.getKey().get(updateCondIdx));
+          updateCaseArgs.add(curUpdateColExprList.get(updateCondIdx));
+        }
+        // Append NULL for the else case
+        updateCaseArgs.add(colNullLiteral);
```

Comment:
Minor nitpick. I think no else is equivalent to ELSE NULL. I'll defer to you on code quality.

Reply:
In sql code, yes. For the rexnode, the else is required.